Detectors/TPC/workflow/include/TPCWorkflow/TPCSectorCompletionPolicy.h

```diff
@@ -0,0 +1,164 @@
+/// the tracker, DPL will synchronize complete data sets from the input routes. Multiple O2 messages
+/// can be received on one input route, a custom completion policy is needed to define the complete
+/// data set. All TPC data except raw data is sent with a specific TPCSectorHeader on the header
+/// stack describing the current sector and defining the active sectors in the setup. The completion
+/// policy callback will wait until there is data for all active sectors.
+///
+/// Parameters:
+///   processor name   rexexp to match a name of the processor for which the policy should be applied
+///   input matchers   provided as an argument pack
+///                    Note: it is important to use ConcreteDataTypeMatcher to define input spec with
+///                          wildcard on subSpecification
+/// Usage:
+///   TPCSectorCompletionPolicy("processor-name-regexp", InputSpec{"", ConcreteDataTypeMatcher{"DET", "RAWDATA"}}, ...)();
+///
+class TPCSectorCompletionPolicy
+{
+ public:
+  TPCSectorCompletionPolicy() = delete;
+  template <typename... Args>
+  TPCSectorCompletionPolicy(const char* processorName, Args&&... args)
+    : mProcessorName(processorName), mInputMatchers()
+  {
+    init(std::forward<Args>(args)...);
+  }
+
+  o2::framework::CompletionPolicy operator()()
+  {
+    constexpr static size_t NSectors = o2::tpc::Sector::MAXSECTOR;
+
+    auto matcher = [expression = mProcessorName](DeviceSpec const& device) -> bool {
+      return std::regex_match(device.name.begin(), device.name.end(), std::regex(expression.c_str()));
+    };
+
+    auto callback = [inputMatchers = mInputMatchers](CompletionPolicy::InputSet inputs) -> CompletionPolicy::CompletionOp {
+      auto op = CompletionPolicy::CompletionOp::Wait;
+      std::bitset<NSectors> validSectors = 0;
+      bool haveMatchedInput = false;
+      uint64_t activeSectors = 0;
+      size_t nActiveInputRoutes = 0;
+      size_t nMaxPartsPerRoute = 0;
+      int inputType = -1;
+      for (auto it = inputs.begin(), end = inputs.end(); it != end; ++it) {
+        nMaxPartsPerRoute = it.size() > nMaxPartsPerRoute ? it.size() : nMaxPartsPerRoute;
+        bool haveActivePart = false;
+        for (auto const& ref : it) {
+          if (!DataRefUtils::isValid(ref)) {
+            continue;
+          }
+          haveActivePart = true;
+          auto const* dh = DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+          // check if the O2 message matches on of the input specs to be matched and
+          // if it matches, check for the sector header, retrieve the active sector information
+          // and mark the sector as valid
+          for (size_t idx = 0, end = inputMatchers.size(); idx < end; idx++) {
+            auto const& spec = inputMatchers[idx];
+            if (DataRefUtils::match(ref, spec)) {
+              haveMatchedInput = true;
+              if (inputType == -1) {
```

Comment:
I think here is a small bug. You do not set the input type afterwards, so it will always remain -1.

Reply:
good catch, fixed. Too bad that the compiler does not get this from my comment :-)